### ISSUE TITLE
Celery results are NOT obeying results serializer config in databases

### ISSUE THREAD

**Issue by fxfitz** [celery/celery#3025]
So, we're working on getting all of our tasks results stored in a database backend (for now just postgres, but we're not tied to any particular backend).

Anyway, we've noticed that no matter what we try to do, the results are **always** stored as a pickle object, even though we have the other configurations set properly:

```python
app.conf.update(CELERY_ACCEPT_CONTENT=["json"],
                CELERY_TASK_SERIALIZER="json",
                CELERY_RESULT_SERIALIZER="json")
```

While not being terribly familiar with SQL Alchemy, I do believe we have found the [offending line](https://github.com/celery/celery/blob/master/celery/backends/database/models.py#L33).

Am I understanding this right? To me, it looks like this has been a bug for _years_... is it because we might be the only ones using a database backend? :-P

Thanks!

**Comment by czchen** [comment]
@fxfitz, you are not along. We plan to give MongoDB a try to see if it is is more suitable for our need.

**Comment by fxfitz** [comment]
@czchen That's great to hear! MongoDB appears to work, so good luck!  I spoke with someone on IRC about this last night, and it looks like they want to see a patch for v4.0 first, and then they'll think about backporting to 3.1.

They're obviously concerned about making any breaking changes, but in the meantime I think documentation should be amended to indicate the bug/inconsistency and this issue should be left open and unresolved.

On a side note, we may write a custom backend that will obey CELERY_RESULT_SERIALIZER.  If I do, I'll contribute upstream.

**Comment by auvipy** [comment]
what's the update on the issue? any pr? workaround?

**Comment by jfach** [comment]
@fxfitz have you been able to solve this issue? Currently dealing with this myself using Sqlite.

@auvipy any internal information/updates you can give us?

**Comment by fxfitz** [comment]
Sadly, I moved off of that project and am working on something else now. Sorry!

**Comment by jfach** [comment]
@fxfitz Thanks for the quick response

**Comment by nick4u** [comment]
Guys .. I lost a lot of time trying to figure it out what is wrong with results stored in database as I expected them to be JSON and there were actually storead as Pickle ... This bug/issue is hanging a long time and the least decent thing to do is to state in documentation that even when you configure as recommended:
```
    task_serializer='json',
    accept_content=['json'], 
    result_serializer='json',
```
in-database results will be Pickle objects.
Best regards - great project.

**Comment by thedrow** [comment]
I don't see the offending line anymore.

@nick4u Did you try this on a recent version of Celery?
Also, if you feel like adding a note about this, feel free to issue a PR.

**Comment by adimux** [comment]
The offending line is still there: https://github.com/celery/celery/blob/master/celery/backends/database/models.py#L29
The result column is of type Pickle. It works because the pickle type simply stores anything you pass to it as a BLOB in the DB and restores it when reading from the field.
The question is: should the result be serialized using the result serializer and then be stored in a text field?

**Comment by georgepsarakis** [comment]
I think that if the [result](https://github.com/celery/celery/blob/e6ea4331215c6ac1dbee38868f212c4d48872b36/celery/backends/database/__init__.py#L113) is [encoded](https://github.com/celery/celery/blob/e6ea4331215c6ac1dbee38868f212c4d48872b36/celery/backends/base.py#L292) before being stored, then the selected serializer will be used.

**Comment by adimux** [comment]
Yes, but the question I am asking is whether we should encode it and store in a text field instead of pickle.

**Comment by georgepsarakis** [comment]
The [SQLAlchemy PickleType](https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.PickleType) column definition could be switched to `String` or `Text`. We need to check which one is the best option in terms of backwards compatibility.

**Comment by thedrow** [comment]
But what if we define the serializer to use json or msgpack?

**Comment by adimux** [comment]
@thedrow Possible, but we'd have to switch the column type depending on the serializer and default to Text when there isn't an appropriate column type for the serialization method chosen.
Also, it wouldn't be backwards-compatible, so people who used SQLAlchemy back-end are going to have to stick to `pickle` as a `result_serializer`, which is ok.

Personally, I'd prefer to vary the column type depending on the result serializer json.

**Comment by adimux** [comment]
Another problem is that the `JSON` SQLAlchemy type only works with databases that support JSON types.
For example, JSON-B didn't exist before Mysql 5.7.

**Comment by georgepsarakis** [comment]
@thedrow the suggested change is to modify the backend to serialize the result, before storing it in the database. At the moment, the code relies to `PickleType` to perform the serialization, which creates the observed inconsistency, when different serialization formats are used.

@adimux only `BLOB/TEXT` type columns can be used as far as I understand, serialization formats cannot be fully mapped to RDBMS column types, with JSON perhaps being the only exception to the rule. Also, users may define their own serializer.

**Comment by adimux** [comment]
@georgepsarakis it's true that the safest is to use `TEXT/BLOB` field for example and to simply serialize the data before storing it.
(As a plus, if we use the json result serializer it's going to work even if the field type is JSON in the DB)

**Comment by nick4u** [comment]
as @fxfitz noted
> ... To me, it looks like this has been a bug for years... is it because we might be the only ones using a database backend? ...

I have switched to Redis as backend soon after posting my remark in this thread so it is not as relevant for me. But before I did, for a short time I've been directly using tables celery saved results and even setup custom result table names 
```
database_table_names = {
    'task': 'appX_taskmeta',
    'group': 'appX_groupmeta',
}
```

... and joined to those tables without ever using Result.get() / Result.forget(). Once I've learn it is Pickle object in db I just did pickle.loads() and continued to live on :)
IF this behaviour was embedded in code for years I really think I am not the only one who got this idea (smart or not :) ) of direct access on SQL level.
Personally I would describe this in documentation and introduce some new transient variable to force new behaviour and then depreciate it in few versions.

**Comment by srikiraju** [comment]
Ran into this too. Do you want to punch out a PR for this?

**Comment by thedrow** [comment]
Sure, go ahead.

**Comment by martincapello** [comment]
I just run into this same issue, and I believe https://github.com/celery/celery/pull/2881 should fix this. Does anyone knows when that PR could be merged? Is there anything I can do to make it happen faster?

**Comment by auvipy** [comment]
that is milestoned for 4.4.x release

**Comment by martincapello** [comment]
Alright, I can wait a bit. Please let me know if nobody is working on this, in which case I could try to create a new PR based on the #2881 work. Thank you.

**Comment by martincapello** [comment]
Well, I couldn't resist the need to create a new PR based on #2881. So #5907 is the rebased version.

**Comment by rodionterekhin** [comment]
Still no progress in 2024???

**Comment by thedrow** [comment]
Contributions are welcome.

**Comment by tamersalama** [comment]
Who's here in 2025?

**Comment by WZhijun154** [comment]
hope updates on this issue/bug

**Comment by auvipy** [comment]
I will personally handle this. Very soon.

**Comment by ghirailghiro** [comment]
Hi @auvipy opened #10536, building on the approach from #2881 and #5907.

The bug is that `Task.result`/`TaskSet.result` use `sa.PickleType`, which
pickles/unpickles at the ORM level regardless of `result_serializer`.

The fix i tried to implement is that columns become `sa.LargeBinary` (same DDL as PickleType under the
hood, no migration needed) and the backend now calls encode()/decode()
itself, same idea as #2881/#5907. Old rows are always pickle no matter the
configured serializer, so decoding falls back to pickle.loads() when the
configured serializer can't parse a row.

In #5907  @auvipy's comment about a native JSON column — tried it, doesn't work:
inserting pickle bytes into a `sa.JSON` column raises "Object of type
bytes is not JSON serializable". Since result_serializer can also be
pickle/msgpack/yaml, LargeBinary is the only type that works for all of
them.

I added a couple of tests on the PR for testing backward-compat
